src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Methodologies/MethodologyService.cs

```diff
@@ -211,12 +177,62 @@ await _context.Entry(methodology.ScheduledWithRelease)
             MethodologyUpdateRequest request)
         {
             return await _persistenceHelper
-                .CheckEntityExists<Methodology>(id, HydrateMethodologyForManageMethodologySummaryViewModel)
+                .CheckEntityExists<Methodology>(id, q =>
+                    q.Include(m => m.MethodologyParent))
                 .OnSuccess(methodology => UpdateMethodologyStatus(methodology, request))
                 .OnSuccess(methodology => UpdateMethodologyDetails(methodology, request))
                 .OnSuccess(_ => GetSummary(id));
         }
 
+        private async Task<MethodologySummaryViewModel> BuildMethodologySummaryViewModel(Methodology methodology)
+        {
+            if (!_context.TryReloadEntity(methodology, out var loadedMethodology))
```

Comment:
Hey @benoutram - presumably this is to reload a Methodology after an update or something in order to have the latest and greatest details?  Just need to understand a bit more why we've got this here, as I've not seen us doing this elsewhere?

Reply:
Hi @duncan-at-hiveit ,

I noticed that we've recently started adding code to explicitly load properties closer in terms of lines to where they are used in an attempt to make it a little bit clearer why they are being included, rather than eagerly loading them with a `.Include` in code that can sometimes be a few calls away from where they are needed.

However, if you try and explicitly load properties of an entity that is not attached to the database context then you get a nasty run-time error along the lines of "attempt was made to load navigation property on detached entity".

One example way of reproducing this would be to write an auth handler that explicitly loads a database property on the entity being tested in an auth context, but call that auth handler with an entity loaded from the database using `.AsNoTracking`. Since the entity is not attached to the database context the auth handler will fail at runtime. The `.AsNoTracking` might have been added to try and improve performance, but with unintended consequences. See `ViewPublicationAuthorizationHandler` and `ViewReleaseAuthorizationHandler` which are now using this reloading technique to ensure the entities are not detached to avoid this kind of scenario.

Within the method changing here to build the view model we need to make sure the Publications are hydrated within the Methodology as the view model uses them.

The Methodology passed into this private method sometimes comes from the persistence helper, but sometimes comes from the `MethodologyRepository` dependency of this service.

From the persistence helper, providing that `.AsNoTracking` hasn't been used, it will be fine, but we have less control over a Methodology returned by the `MethodologyRepository`.

I would expect the Methodology returned by `MethodologyRepository` to be fine too, but we see a problem in the tests where this becomes useful to add in.

In the tests for methods that use, e.g. `[Fact] public async Task GetAdoptableMethodologies()`  a Methodology is saved in the database, but the service is run with a fresh database context where the Methodology is returned by a mock. Since it's not attached to the database context, but is present in the database, the `TryReloadEntity` will load it and the Publications can be explicitly loaded.

Another way of solving this would be to use `context.Attach(methodology)` just prior to calling the service in the test but I thought this was a good use of the `TryReloadEntity` now that we have it.

Happy to remove this and use `context.Attach` in the unit tests if you think that would be better. 🙂

Reply:
Hey Ben - fine with it this way, but I'm personally not a fan of the way it looks, so having it scattered around the codebase, it might be a bit ugly as we use it more often.  I wonder if the next time we use it we could consider wrapping it in an extension method of our own that would do the check and throw a consistent, useful error message if it could not be reloaded, and all as a one-liner?

Reply:
Or something that returns an Either...

Reply:
I agree, I think we need a variation of `ReloadEntity` and `TryReloadEntity` that ensures it's loaded without returning an entity that can be null.

Reply:
`AssertEntityLoaded`?

Reply:
Just added this in @ntsim @duncan-at-hiveit 

```
        /// <summary>
        /// Ensure that an entity is not detached from the database context, reloading it if necessary before returning.
        /// Throws an exception if the entity cannot be reloaded.
        /// </summary>
        public static T AssertEntityLoaded<T>(this DbContext context, T entity) where T : class
        {
            var reloaded = context.ReloadEntity(entity);

            if (reloaded is null)
            {
                var displayString = context.GetDisplayString(entity);
                throw new InvalidOperationException($"Unable to reload {displayString}");
            }

            return reloaded;
        }
```

```
        private static string GetDisplayString(this DbContext context, object entity)
        {
            var entry = context.Entry(entity);
            var primaryKey = entry.Metadata.FindPrimaryKey();
            var typeName = entity.GetType().ShortDisplayName();
            var primaryKeyValues = primaryKey.Properties
                .Select(property => $"{property.Name}: '{property.PropertyInfo.GetValue(entity)}'")
                .JoinToString(", ");

            return $"{typeName} {{{primaryKeyValues}}}";
        }
```

which lets me replace the ugly code with:

`var loadedMethodology = _context.AssertEntityLoaded(methodology);`

Reply:
Very nice :grin: